Sources/X509/Verifier/RFC5280/IPConstraints.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftASN1
+
+extension NameConstraintsPolicy {
+    /// Validates that an IP address matches a constraint.
+    ///
+    /// The rules for IP address constraints are fairly simple. The constraint contains both a subnet
+    /// and a subnet mask, while the `ipAddress` will contain just the bytes of the address. A constraint
+    /// matches if the address is part of the subnet defined by the mask.
+    ///
+    /// Additionally, RFC 5280 requires that the constraint be equivalent to a subnet defined using CIDR notation.
+    /// This implies that we do not tolerate arbitrary masks.
+    @inlinable
+    static func ipAddressMatchesConstraint(ipAddress: ASN1OctetString, constraint: ASN1OctetString) -> Bool {
+        switch (ipAddress.bytes.count, constraint.bytes.count) {
+        case (4, 8):
+            // IPv4
+            return addressIsInSubnet(address: ipAddress.bytes, subnet: constraint.bytes)
+        case (16, 32):
+            // IPv6
+            return addressIsInSubnet(address: ipAddress.bytes, subnet: constraint.bytes)
+        default:
+            // No match or an invalid format.
+            return false
+        }
+    }
+}
+
+extension ArraySlice where Element == UInt8 {
+    @inlinable
+    var isValidCIDRMask: Bool {
```

Comment:
I'm wondering what I general rule of thumb is on when we want to add functionally as an extension like this or declare it to as a free or static function.

I personally like these extensions, it's just a pity that we can't make them `fileprivate`. For that reason I often fall back to static functions to restrict its visibility to a related namespace.

Reply:
Yeah, I'm largely not too worried about it here, but I get the concern. For larger projects I think it would be a bigger deal, but in this instance I think we're probably safe.